space-invaders.asm

```diff
@@ -7,6 +7,27 @@
 %define INVADERS_SHOOT_CYCLES 4
 %define BULLETS_MOVE_CYCLE 28
 
+%define INVALID_STATE 0x0000
+
+%define PLAYER_WIN 0
+%define INVADERS_WIN 1
+
+%define START_STATE 0
+%define GAME_STATE 1
+%define END_STATE 2
+
+%define START_KEY ' '
+%define RETRY_KEY 'r'
+
+%define MOVE_UP 0
+%define MOVE_RIGHT 1
+%define MOVE_DOWN 2
+%define MOVE_LEFT 3
+
+%define INVADERS_NEXT_OFFSET 2
+%define BULLET_POSITION_OFFSET 1
```

Comment:
Nice Idea. However, I would go all in on this approach and completely define the bullet "struct" to improve concistency. Additionally, I would import the corresponsing bullet status constants to this location as well. So something like:

```asm
%define BULLET_STATUS_OFFSET 0
%define BULLET_STATUS_SIZE 1
%define BULLET_POSITION_OFFSET BULLET_STATUS_OFFSET + BULLET_STATUS_SIZE
%define BULLET_POSITION_SIZE 2
%define BULLET_SIZE BULLET_POSITION_OFFSET + BULLET_POSITION_SIZE

%define BULLET_STATUS_PLAYER 'p'
%define BULLET_STATUS_INVADER 'i'
```

Reply:
Oh, that's good idea!
Then it makes it easier to understand the data structure of bullet "struct"!
I will fix it soon.